Swagger Codegen can't use my own template files

Hello

When  I set template directory for swagger-codegen-maven-plugin to generate java code from my own template files.
the build fails.
I use this configuration for maven plugin:

                <plugin>
                    <groupId>io.swagger.codegen.v3</groupId>
                    <artifactId>swagger-codegen-maven-plugin</artifactId>
                    <version>3.0.30</version>
                    <executions>
                        <execution>
                            <goals>
                                <goal>generate</goal>
                            </goals>
                            <configuration>
                                <inputSpec>${project.basedir}/src/main/resources/openapi.json</inputSpec>
                                <language>spring</language>
                                <library>spring-cloud</library>
                                <templateDirectory>src/main/resources/JavaSpring</templateDirectory>
                                <configOptions>
                                    <generateForOpenFeign>true</generateForOpenFeign>
                                    <groupId>${project.parent.groupId}</groupId>
                                    <artifactId>${project.name}-java-client</artifactId>
                                    <artifactVersion>${project.version}</artifactVersion>
                                    <dateLibrary>java8</dateLibrary>
                                </configOptions>
                            </configuration>
                        </execution>
                    </executions>
                </plugin>

The plugin can't find my files :
java.io.FileNotFoundException: /handlebars/JavaSpring/C:/Users/myname/Desktop/projects/test-project-api/src/main/resources/JavaSpring/model.mustache

I was wondering if the problem caused by wrong prefix of URL ?

I'm facing the same issue using the CLI. It seems the prefix `/handlebars/<lang>` is always added regardless of indicating the templateDir path. Note `Caused by: java.io.FileNotFoundException: /handlebars/csharp/C:/source/oas3-migration/dotnet-swagger-templates/model.mustache` in the following error log:

```
java --add-opens=java.base/java.util=ALL-UNNAMED -jar modules/swagger-codegen-cli/target/swagger-codegen-cli.jar generate -l csharp -i C:/spec.json -o /tmp/dotnet -t C:\source\oas3-migration\dotnet-swagger-templates

15:00:42.525 [Thread-0] INFO  i.s.c.v.g.dotnet.CSharpClientCodegen - Generating code for .NET Framework v4.5
Exception in thread "Thread-0" java.lang.RuntimeException: Could not generate model 'AccessPolicy'
        at io.swagger.codegen.v3.DefaultGenerator.generateModels(DefaultGenerator.java:451)
        at io.swagger.codegen.v3.DefaultGenerator.generate(DefaultGenerator.java:779)
        at io.swagger.codegen.v3.cli.cmd.Generate.run(Generate.java:388)
        at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: java.io.FileNotFoundException: /handlebars/csharp/C:/source/oas3-migration/dotnet-swagger-templates/model.mustache
        at com.github.jknack.handlebars.io.URLTemplateLoader.sourceAt(URLTemplateLoader.java:70)
        at com.github.jknack.handlebars.Handlebars.compile(Handlebars.java:438)
        at com.github.jknack.handlebars.Handlebars.compile(Handlebars.java:419)
        at io.swagger.codegen.v3.templates.HandlebarTemplateEngine.getHandlebars(HandlebarTemplateEngine.java:44)
        at io.swagger.codegen.v3.templates.HandlebarTemplateEngine.getRendered(HandlebarTemplateEngine.java:24)
        at io.swagger.codegen.v3.DefaultGenerator.processTemplateToFile(DefaultGenerator.java:795)
        at io.swagger.codegen.v3.DefaultGenerator.generateModels(DefaultGenerator.java:438)
        ... 3 more
```

--- a/modules/swagger-codegen/src/main/java/io/swagger/codegen/v3/templates/CodegenTemplateLoader.java
+++ b/modules/swagger-codegen/src/main/java/io/swagger/codegen/v3/templates/CodegenTemplateLoader.java
@@ -1,39 +1,54 @@
 package io.swagger.codegen.v3.templates;
 
-import com.github.jknack.handlebars.io.FileTemplateLoader;
-import org.apache.commons.lang3.StringUtils;
+import com.github.jknack.handlebars.io.URLTemplateLoader;
 
 import java.io.File;
 import java.io.IOException;
 import java.net.URL;
 
-public class CodegenTemplateLoader extends FileTemplateLoader {
+public class CodegenTemplateLoader extends URLTemplateLoader {
     private String templateDir;
     private String customTemplateDir;
 
-    public CodegenTemplateLoader(String templateDir, String suffix) {
-        super(templateDir, suffix);
-        this.templateDir = templateDir;
+    public CodegenTemplateLoader() {
+        setSuffix(".mustache");
     }
 
+    @Override
+    public String resolve(String uri) {
+        if (!uri.endsWith(this.getSuffix())) {
+            uri = uri + this.getSuffix();
+        }
+        File templateFile = new File(uri);
+        if (templateFile.exists()) {
+            return templateFile.toString();
+        }
+        templateFile = new File(this.getPrefix() + this.normalize(uri));
+        if (templateFile.exists()) {
+            return templateFile.toString();
+        }
+        if (this.customTemplateDir != null) {
+            templateFile = new File(this.customTemplateDir, this.normalize(uri));
+            if (templateFile.exists()) {
+                return templateFile.toString();
+            }
+        }
+        if (getClass().getResource(this.getPrefix() + this.normalize(uri)) != null) {
+            return this.getPrefix() + this.normalize(uri);
+        }
+        return this.templateDir + this.normalize(uri);
+    }
+
+    @Override
     public URL getResource(String location) throws IOException {
         if (this.customTemplateDir == null) {
             return this.getClass().getResource(location);
         }
-        String templateFile = resolveTemplateFile(this.templateDir, location);
-        templateFile = resolveTemplateFile(this.customTemplateDir, templateFile);
-        final File file = new File(this.customTemplateDir, templateFile);
+        final File file = new File(location);
         if (file.exists()) {
             return file.toURI().toURL();
         }
         return this.getClass().getResource(location);
-    }
-
-    private String resolveTemplateFile(String templateDir, String templateFile) {
-        if (templateFile.startsWith(templateDir)) {
-            templateFile = StringUtils.replaceOnce(templateFile, templateDir, StringUtils.EMPTY);
-        }
-        return templateFile;
     }
 
     public String getCustomTemplateDir() {
@@ -46,6 +61,22 @@ public class CodegenTemplateLoader extends FileTemplateLoader {
 
     public CodegenTemplateLoader customTemplateDir(String customTemplateDir) {
         this.customTemplateDir = customTemplateDir;
+        return this;
+    }
+
+    public String getTemplateDir() {
+        return templateDir;
+    }
+
+    public void setTemplateDir(String templateDir) {
+        this.templateDir = this.getPrefix() + templateDir;
+        if (!this.templateDir.endsWith("/")) {
+            this.templateDir = this.templateDir + "/";
+        }
+    }
+
+    public CodegenTemplateLoader templateDir(String templateDir) {
+        setTemplateDir(templateDir);
         return this;
     }
 }

--- a/modules/swagger-codegen/src/main/java/io/swagger/codegen/v3/templates/HandlebarTemplateEngine.java
+++ b/modules/swagger-codegen/src/main/java/io/swagger/codegen/v3/templates/HandlebarTemplateEngine.java
@@ -1,13 +1,10 @@
 package io.swagger.codegen.v3.templates;
 
 import com.github.jknack.handlebars.Handlebars;
-import com.github.jknack.handlebars.io.ClassPathTemplateLoader;
 import com.github.jknack.handlebars.io.TemplateLoader;
 import io.swagger.codegen.v3.CodegenConfig;
 import io.swagger.codegen.v3.CodegenConstants;
-import org.apache.commons.lang3.StringUtils;
 
-import java.io.File;
 import java.io.IOException;
 import java.util.Map;
 
@@ -31,23 +28,16 @@ public class HandlebarTemplateEngine implements TemplateEngine {
     }
 
     private com.github.jknack.handlebars.Template getHandlebars(String templateFile) throws IOException {
-        templateFile = templateFile.replace(".mustache", StringUtils.EMPTY).replace("\\", "/");
-        String templateDir = config.customTemplateDir() != null ? "" : config.templateDir().replace("\\", "/");
-        String customTemplateDir = config.customTemplateDir() != null ? config.customTemplateDir().replace("\\", "/") : null;
+        templateFile = templateFile.replace("\\", "/");
+        final String templateDir = config.templateDir().replace("\\", "/");
         final TemplateLoader templateLoader;
-        templateFile = resolveTemplateFile(templateDir, templateFile);
-        templateLoader = new CodegenTemplateLoader("/" + templateDir, ".mustache")
+        String customTemplateDir = config.customTemplateDir() != null ? config.customTemplateDir().replace("\\", "/") : null;
+        templateLoader = new CodegenTemplateLoader()
+                .templateDir(templateDir)
                 .customTemplateDir(customTemplateDir);
         final Handlebars handlebars = new Handlebars(templateLoader);
         handlebars.prettyPrint(true);
         config.addHandlebarHelpers(handlebars);
         return handlebars.compile(templateFile);
-    }
-
-    private String resolveTemplateFile(String templateDir, String templateFile) {
-        if (templateFile.startsWith(templateDir)) {
-            templateFile = StringUtils.replaceOnce(templateFile, templateDir, StringUtils.EMPTY);
-        }
-        return templateFile;
     }
 }

